### PR TITLE
Support promotion to more than two regions

### PR DESCRIPTION
**Pull request by boumenot**
The CLI had three commands for promotion.

 * promote-single-region
 * promote-two-regions
 * promote-all-regions

In truth, we had much more flexibility than this, but the current interface is limiting.  I collapsed promote-single-region and promote-two-regions into the command promote.  The user can pass an arbitrary number of regions to promote an extension to.

The CLI used to a do a search and replace of the key string %REGIONS%, which was expedient, but not correct.  This update handles adding a Regions element to the document as necessary.  Order of the document is important to RDFE, which this code is now able to handle.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -61,14 +61,10 @@ var (
...
-		Name:   "region-1",
-		Usage:  "Primary pilot location to roll out the extension (e.g. 'Japan East')",
-		EnvVar: "REGION1"}
-	flRegion2 = cli.StringFlag{
-		Name:   "region-2",
-		Usage:  "Primary pilot location to roll out the extension (e.g. 'Brazil South')",
-		EnvVar: "REGION2"}
+	flRegion = cli.StringSliceFlag{
+		Name:  "region",
+		Usage: "List of one or more regions to rollout an extension (e.g. 'Japan East')",
```
Comment (hglkrijger) [inline main.go]: Maybe give an example of multiple regions as well

File: manifest.go [hunk 2]
```diff
@@ -1,76 +1,175 @@
...
 
+type certificate struct {
+	StoreLocation       string `xml:"StoreLocation,omitempty"`
+	StoreName           string `xml:"StoreName,omitempty"`
+	ThumbprintRequired  bool   `xml:"ThumbprintRequired,omitempty"`
+	ThumbprintAlgorithm string `xml:"ThumbprintAlgorithm,omitempty"`
+}
+
+// NOTE(@boumenot): there is probably a better way to express this.  If
+// you know please share...
```
Comment (hglkrijger) [inline manifest.go]: @brendandixon any thoughts on this?

File: manifest.go [hunk 3]
```diff
@@ -1,76 +1,175 @@
...
+	SupportedOS                string       `xml:"SupportedOS,omitempty"`
+	Regions                    string       `xml:"Regions"`
+}
+
+type extensionManifest interface {
+	Marshal() ([]byte, error)
+}
+
+func isGuestAgent(providerNameSpace string) bool {
+	return "Microsoft.OSTCLinuxAgent" == providerNameSpace
```
Comment (hglkrijger) [inline manifest.go]: I wonder if we should add the Windows agent too.

File: manifest.go [hunk 4]
```diff
@@ -1,76 +1,175 @@
...
-  <CompanyName>{{.Company}}</CompanyName>
-  <SupportedOS>{{.OS}}</SupportedOS>
-  <!--%REGIONS%-->
-</ExtensionImage>
-`
-	tpl, err := template.New("manifest").Parse(manifestXML)
-	if err != nil {
-		log.Fatalf("template parse error: %v", err)
+
+	manifest := extensionImage{
```
Comment (hglkrijger) [inline manifest.go]: Much cleaner

File: manifest_test.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+import (
+	"bytes"
+	"encoding/xml"
+	"testing"
+
+	"github.com/approvals/go-approval-tests"
+)
+
+func TestRoundTripExtensionImage(t *testing.T) {
+	xmlString := []byte(`<ExtensionImage xmlns="http://schemas.microsoft.com/windowsazure">
```
Comment (hglkrijger) [inline manifest_test.go]: Can these reference the txt files?

File: vendor/github.com/approvals/go-approval-tests/.travis.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+install:
+  - go get -u github.com/golang/lint/golint
+
+script:
+  - test -z "$(go fmt -s -l -w ./... | tee /dev/stderr)"
+  - golint -set_exit_status ./...
+  - go vet ./...
+  - go build -v ./...
+  - go test -v ./...
```
Comment (hglkrijger) [inline vendor/github.com/approvals/go-approval-tests/.travis.yml]: Not sure if you need a newline for yml

File: main.go (hunk 1 above)
Comment (boumenot) [inline main.go]: I added a section to the README.md to better demonstrate this.

File: manifest.go (hunk 3 above)
Comment (boumenot) [inline manifest.go]: It is a different deployment mechanism, and I don't think it is going to change. ☹️

File: vendor/github.com/approvals/go-approval-tests/.travis.yml (hunk 6 above)
Comment (boumenot) [inline vendor/github.com/approvals/go-approval-tests/.travis.yml]: Nope.

File: manifest_test.go (hunk 5 above)
Comment (boumenot) [inline manifest_test.go]: Yes, but I tend to prefer them inline.  What's your thinking?

**Pull request merged**